DesignSafe-CI/portal
# Task/DES-1547 - File Metadata

## Overview: ##
**11/3/21 - Update**: I'll need to make a small change to the requirements file before we merge this PR. We'll need to request an update from the agavepy folks so we can set the latest version of agavepy with bulk operation support in our requirements file. Additionally, we should double check on the file metadata namespace being used in production. I don't think there is anything in the database seriously using `designsafe.file`, (there's only like 200ish records mostly with test information in them) but it's just something to consider.

This is a PR to test updated agave file operations which will move, copy, and create metadata associated with individual files. In this PR you can manage file metadata using new agavepy methods as well as the utils in `designsafe/apps/api/datafiels/utils.py`. You will be able to see a file's metadata in the metadata header of the preview modal.

Things to keep in mind:
- File metadata does not have granular permissions. File operations are carried out if the user is authorized, and (if authorized) metadata will be modified as needed by the service_account in the same operation.
- Most images and raw file types will automatically have their metadata scraped and added into the system. This will happen once the files are uploaded. One issue to note is that some of these metadata fields can contain non-human-readable data.
- The iterator for duplicated file names has been changed from a date format to a numerical format (similar to Frontera).
- When files are "deleted" (moved to the .Trash directory) we may need to make a task that will purge any metadata in the .Trash directories of private and project systems.
- The `file_meta_obj` function makes a default object for file metadata creation. This could probably be a class object if we wanted to do something fancy with project related file tags later on, but it might be overkill.
- All file metadata is named `designsafe.file` in our database, and can be queried by `path` and `system`.


## PR Status: ##


## Related Jira tickets: ##

* [DES-1547](https://jira.tacc.utexas.edu/browse/DES-1547)

## Testing Steps: ##
1. `npm ci` and `make build`
2.  upload some sort of image file or raw file (if you need one for testing poke me on slack)
3. you should be able to see new metadata when previewing the file
4. try to copy/move the file to different places using designsafe. The metadata should follow it if moved or copied. You should also be able to copy/move a directory containing the file and see the same thing.
